File: orchard_fem/solvers/static_preload.py
from __future__ import annotations

from typing import TYPE_CHECKING

from orchard_fem.solvers._petsc import create_aij_matrix, require_petsc, solve_linear_system


if TYPE_CHECKING:
    from orchard_fem.solvers.modal_assembler import LinearDynamicAssemblyResult


def _mat_vec(matrix: list[list[float]], vector: list[float]) -> list[float]:
    return [
        sum(row[column] * vector[column] for column in range(len(vector)))
        for row in matrix
    ]


def _local_displacement(
    transformation_matrix: list[list[float]],
    global_displacement: list[float],
    dofs: tuple[int, ...],
) -> list[float]:
    element_global = [global_displacement[dof] for dof in dofs]
    return [
        sum(transformation_matrix[row][column] * element_global[column] for column in range(len(dofs)))
        for row in range(len(dofs))
    ]


def compute_gravity_axial_forces(
    assembled: "LinearDynamicAssemblyResult",
    gravity_load: list[float],
) -> dict[str, list[float]]:
    """
    Solve the static gravity preload and recover element axial forces.
    """
    require_petsc()

    if len(gravity_load) != len(assembled.dof_labels):
        raise ValueError("gravity_load size must match the assembled DOF count")

    stiffness = create_aij_matrix(assembled.stiffness_matrix)
    static_displacement = solve_linear_system(stiffness, gravity_load)
    residual = _mat_vec(assembled.stiffness_matrix, static_displacement)
    residual_norm = max(
        (abs(expected - actual) for expected, actual in zip(gravity_load, residual)),
        default=0.0,
    )
    if residual_norm > 1.0e-5:
        raise RuntimeError(f"Gravity preload solve residual is too large: {residual_norm}")

    axial_forces: dict[str, list[float]] = {}
    for branch_id, elements in assembled.branch_elements.items():
        branch_forces: list[float] = []
        for element in elements:
            local_displacement = _local_displacement(
                element.transformation_matrix,
                static_displacement,
                element.dofs,
            )
            axial_force = element.axial_rigidity * (
                (local_displacement[6] - local_displacement[0]) / element.length
            )
            branch_forces.append(axial_force)
        axial_forces[branch_id] = branch_forces

    return axial_forces

File: src/solver_core/StaticPreload.cpp
#include "orchard_solver/solver_core/StaticPreload.h"

#include <algorithm>
#include <cmath>
#include <stdexcept>
#include <string>

#include "orchard_solver/solver_core/LinearAlgebra.h"

namespace orchard {

namespace {

std::vector<std::vector<double>> copyDenseMatrix(const DenseMatrix& matrix) {
    std::vector<std::vector<double>> result(
        matrix.rows(),
        std::vector<double>(matrix.cols(), 0.0)
    );
    for (std::size_t row = 0; row < matrix.rows(); ++row) {
        for (std::size_t col = 0; col < matrix.cols(); ++col) {
            result[row][col] = matrix(row, col);
        }
    }
    return result;
}

std::vector<double> multiply(const DenseMatrix& matrix, const std::vector<double>& vector) {
    std::vector<double> result(matrix.rows(), 0.0);
    for (std::size_t row = 0; row < matrix.rows(); ++row) {
        for (std::size_t col = 0; col < matrix.cols(); ++col) {
            result[row] += matrix(row, col) * vector[col];
        }
    }
    return result;
}

std::array<double, 12> localDisplacement(
    const Matrix12& transformation,
    const std::vector<double>& displacement,
    const std::array<int, 12>& dofs
) {
    std::array<double, 12> global_element {};
    for (std::size_t index = 0; index < dofs.size(); ++index) {
        global_element[index] = displacement.at(static_cast<std::size_t>(dofs[index]));
    }

    std::array<double, 12> local {};
    for (std::size_t row = 0; row < 12; ++row) {
        for (std::size_t col = 0; col < 12; ++col) {
            local[row] += transformation[row][col] * global_element[col];
        }
    }
    return local;
}

} // namespace

std::unordered_map<std::string, std::vector<double>> computeGravityAxialForces(
    const AssembledModel& assembled,
    const std::vector<double>& gravity_load
) {
    if (gravity_load.size() != assembled.system.dof_labels.size()) {
        throw std::runtime_error("gravity_load size must match the assembled DOF count");
    }

    const auto stiffness = copyDenseMatrix(assembled.system.stiffness);
    const auto displacement = solveLinearSystem(stiffness, gravity_load);
    const auto residual = multiply(assembled.system.stiffness, displacement);

    double residual_norm = 0.0;
    for (std::size_t index = 0; index < gravity_load.size(); ++index) {
        residual_norm = std::max(residual_norm, std::abs(gravity_load[index] - residual[index]));
    }
    if (residual_norm > 1.0e-5) {
        throw std::runtime_error("Gravity preload solve residual is too large: " + std::to_string(residual_norm));
    }

    std::unordered_map<std::string, std::vector<double>> axial_forces;
    for (const auto& [branch_id, elements] : assembled.branch_elements) {
        auto& branch_forces = axial_forces[branch_id];
        branch_forces.reserve(elements.size());

        for (const auto& element : elements) {
            const auto local = localDisplacement(
                element.transformation,
                displacement,
                element.dofs
            );
            branch_forces.push_back(
                element.axial_rigidity * ((local[6] - local[0]) / element.length)
            );
        }
    }

    return axial_forces;
}

} // namespace orchard
